<a href="https://colab.research.google.com/github/unatardedemartes-a11y/ClassFiles/blob/main/Actividad_Calificada_07_Datos_Temporales_275334.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos temporales**
### Limpieza y transformación de fechas, horas y duraciones en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: datos temporales |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Viajes de taxi en Nueva York, marzo de 2019 (NYC TLC, muestra publicada en un repositorio público de GitHub) |
| **Fecha de entrega** | 26-Sep-2026 |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |Livia Daniela Padilla Barragan |
| **Matrícula** |275334 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en tus resultados (cita cifras).
7. Usa los métodos vistos en la sesión: `pd.to_datetime()`, el accesor `.dt`, `pd.Timedelta`, `.between()`, `tz_localize()`, `groupby()`, `sort_values()`, `.where()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Temporales_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset contiene 6,433 viajes de taxis amarillos y verdes de la ciudad de Nueva York registrados en **marzo de 2019**. Cada fila es un viaje. Las fechas y horas están en **hora local de Nueva York**.

| Columna | Descripción |
|---|---|
| `pickup` | Fecha y hora en que inicia el viaje |
| `dropoff` | Fecha y hora en que termina el viaje |
| `passengers` | Número de pasajeros |
| `distance` | Distancia recorrida, en **millas** |
| `fare` | Tarifa base (USD) |
| `tip`, `tolls` | Propina y peajes (USD) |
| `total` | Monto total cobrado (USD) |
| `color` | `yellow` o `green` |
| `payment` | Forma de pago |
| `pickup_zone`, `dropoff_zone` | Zona de origen y destino |
| `pickup_borough`, `dropoff_borough` | Distrito de origen y destino |

> **Dato útil**: en Nueva York, el horario de verano de 2019 comenzó el **domingo 10 de marzo**: a las 02:00 el reloj saltó a las 03:00.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Conversión y validación** | 20 | 18–20: conversión con formato explícito y verificación de `NaT`; viajes fuera de periodo detectados; análisis del cambio de horario correcto; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o conclusión. | 1–13: conversión sin verificar o validaciones incompletas; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Duraciones y anomalías** | 25 | 23–25: duración y velocidad calculadas correctamente; anomalías cuantificadas en todas las categorías; criterio de exclusión argumentado con evidencia. | 18–22: cálculos correctos; criterio parcial o con poca evidencia. | 1–17: cálculos con errores o anomalías no cuantificadas. | 0: no se entregó o es incorrecto. |
| **3. Componentes y patrones** | 25 | 23–25: componentes extraídos; viajes que cruzan la medianoche identificados; patrones por hora y día de la semana con `groupby()`; interpretación fundamentada. | 18–22: componentes y agrupaciones correctos; interpretación incompleta. | 1–17: agrupaciones con errores o sin interpretación. | 0: no se entregó o es incorrecto. |
| **4. Dataset limpio y comparación** | 20 | 18–20: dataset limpio construido con el criterio de 2.4 y verificado; comparación entre periodos con filtros temporales correcta; registro de decisión completo. | 14–17: dataset limpio correcto; comparación o registro incompletos. | 1–13: criterio aplicado de forma inconsistente o sin verificación. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Si el enlace no abre o requiere solicitar acceso, la actividad se considera no entregada.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [40]:
import pandas as pd

# Carga del dataset de taxis de Nueva York desde un repositorio público de GitHub.
# El archivo se lee con pandas; no se requiere ninguna otra biblioteca.
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv'
taxis = pd.read_csv(url)

print('Dimensiones:', taxis.shape)
taxis.head()

Dimensiones: (6433, 14)


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


---
## Actividad 1. Conversión y validación (20 puntos)

### 1.1 Conversión (5 pts)

1. Muestra el tipo de dato de `pickup` y `dropoff`.
2. Convierte ambas columnas a tipo fecha con `pd.to_datetime()`, indicando el **formato explícito** que corresponde a los valores.
3. Verifica que la conversión no generó `NaT`: compara el número de faltantes antes y después.

In [41]:
# Escribe tu código aquí
import  datetime as dt  #importamos datetime de pandas

#verificar cuantos registro con informacion faltante tienen ambas columnas
faltantes_antes = taxis[['pickup', 'dropoff']].isna().sum()
print("Faltantes antes:", faltantes_antes)

#mostrar tipo de dato de ambas columnas
print()
print("Tipo de dato de columna 'pickup' :",taxis['pickup'].dtype)
print("Tipo de dato de columna 'dropoff':",taxis['dropoff'].dtype)

#convertir datos a datime dado que la columna contiene fecha y hora
taxis['pickup']=pd.to_datetime(taxis['pickup'],format='%Y-%m-%d %H:%M:%S')
taxis['dropoff']=pd.to_datetime(taxis['dropoff'],format='%Y-%m-%d %H:%M:%S')

#verificar cuantos registro con informacion faltante tienen ambas columnas despues de la conversion
print()
faltantes_despues = taxis[['pickup', 'dropoff']].isna().sum()
print("Faltantes después:", faltantes_despues)

Faltantes antes: pickup     0
dropoff    0
dtype: int64

Tipo de dato de columna 'pickup' : object
Tipo de dato de columna 'dropoff': object

Faltantes después: pickup     0
dropoff    0
dtype: int64


### 1.2 Periodo de los datos (5 pts)

El dataset debería contener solo viajes de marzo de 2019.

1. Muestra la fecha mínima y máxima de `pickup` y de `dropoff`.
2. Cuenta y muestra los viajes cuyo `pickup` **no** está en marzo de 2019.
3. ¿Hay viajes con `pickup` en marzo y `dropoff` fuera de marzo? Cuéntalos.

In [42]:
# Escribe tu código aquí

#mostrar fecha maxima y minima para cada columna
print("Fecha mínima de 'pickup':", taxis['pickup'].min())
print("Fecha máxima de 'pickup':", taxis['pickup'].max())
print("Fecha mínima de 'dropoff':", taxis['dropoff'].min())
print("Fecha máxima de 'dropoff':", taxis['dropoff'].max())

#contar cuantos viajes no son de marzo 2019 para columna pickup
pickup_marzo_2019 = (taxis['pickup'].dt.year == 2019) & (taxis['pickup'].dt.month == 3)
fuera_rango = (~pickup_marzo_2019).sum()  #contamos los registros cuyo valor del mes y anio de pickup date no sea marzo 2019
print("Viajes fuera de rango para 'pickup':", fuera_rango)

#contar cuantos viajes empezaron en marzo y terminaron en otro mes
dropoff_marzo_2019 = (taxis['dropoff'].dt.year == 2019) & (taxis['dropoff'].dt.month == 3) #viajes que terminaron en marzo para posteriormente restarlos y obtener solo viajes que inician en marzo y terminan en otro mes
casos = (pickup_marzo_2019 & ~dropoff_marzo_2019).sum()
print("Viajes con pickup en marzo y dropoff fuera de marzo:", casos)




Fecha mínima de 'pickup': 2019-02-28 23:29:03
Fecha máxima de 'pickup': 2019-03-31 23:43:45
Fecha mínima de 'dropoff': 2019-02-28 23:32:35
Fecha máxima de 'dropoff': 2019-04-01 00:13:58
Viajes fuera de rango para 'pickup': 1
Viajes con pickup en marzo y dropoff fuera de marzo: 1


### 1.3 Cambio de horario (5 pts)

El 10 de marzo de 2019, en Nueva York, las horas entre 02:00 y 02:59 **no existieron**.

1. Filtra los viajes con `pickup` el 10 de marzo y cuenta cuántos hay en cada hora (`.dt.hour` y `groupby()` o `value_counts()`). ¿Hay viajes a las 2:00?
2. Localiza `pickup` en la zona `'America/New_York'` con `tz_localize()` usando `nonexistent='NaT'` y cuenta cuántos valores quedan como `NaT`.

In [43]:
# Escribe tu código aquí

#filtrar viajes del 10 de marzo de 2019
viajes_10marzo = taxis[taxis['pickup'].dt.date == pd.Timestamp('2019-03-10').date()]

#conteo de viajes a las 2 am
hora_2 = (viajes_10marzo['pickup'].dt.hour == 2).sum()
print("Viajes con pickup a las 2:00 el 10 de marzo:", hora_2)


#localizar a America/New_York con nonexistent='NaT'
pickup_ny = taxis['pickup'].dt.tz_localize(
    'America/New_York',
    nonexistent='NaT'
)

nat_count = pickup_ny.isna().sum()
print()
print("Valores que quedaron como NaT:",nat_count)


Viajes con pickup a las 2:00 el 10 de marzo: 0

Valores que quedaron como NaT: 0


### 1.4 Conclusión (5 pts)

¿Qué problemas temporales encontraste en esta actividad? ¿Qué indica el resultado de 1.3 sobre la zona horaria en que están registrados los datos? Responde con cifras.

**Respuesta 1.4:**

Las columnas de tiempo estaban en formato de tipo objeto y se tuvo que convertir a tipo datetime y el no haya viajes a las 2:00 y que tampoco haya valores como 'NaT' indica que los datos no están en hora local de Nueva York

---
## Actividad 2. Duraciones y anomalías (25 puntos)

### 2.1 Duración de cada viaje (6 pts)

1. Crea la columna `duracion` = `dropoff` − `pickup`.
2. Crea `duracion_min` con la duración en **minutos** (usa `.dt.total_seconds()`).
3. Muestra el resumen estadístico de `duracion_min` con `describe()`.

In [44]:
# Escribe tu código aquí

#crear la columna 'duracion' como diferencia entre dropoff y pickup
taxis['duracion'] = taxis['dropoff'] - taxis['pickup']

#convertir la duracion a minutos
taxis['duracion_min'] = taxis['duracion'].dt.total_seconds() / 60

#resumen estadistico de duracion_min
print(taxis['duracion_min'].describe())

count    6433.000000
mean       14.349617
std        11.643892
min         0.000000
25%         6.500000
50%        10.900000
75%        18.516667
max       107.666667
Name: duracion_min, dtype: float64


### 2.2 Duraciones anómalas (6 pts)

Cuenta los viajes con:

- Duración **negativa** (termina antes de iniciar).
- Duración **cero**.
- Duración **menor a 1 minuto** (incluye los de duración cero).
- Duración **mayor a 3 horas**.

Muestra los viajes de duración cero con las columnas `pickup`, `dropoff`, `distance`, `fare` y `total`.

In [45]:
# Escribe tu código aquí
#duracion negativa
negativas = (taxis['duracion_min'] < 0).sum()
print("Viajes con duracion negativa:", negativas)

#duracion cero
cero = (taxis['duracion_min'] == 0).sum()
print("Viajes con duracion cero:", cero)

#duracion menor a 1 minuto
menor_1min = (taxis['duracion_min'] < 1).sum()
print("Viajes con duracion menor a 1 minuto:", menor_1min)

#duracion mayor a 3 horas
mayor_3h = (taxis['duracion_min'] > 180).sum()
print("Viajes con duracion mayor a 3 horas:", mayor_3h)

#mostrar viajes de duracion cero con columnas específicas
cero_df = taxis[taxis['duracion_min'] == 0][
    ['pickup', 'dropoff', 'distance', 'fare', 'total']
]
print()
print("Viajes con duración cero")
print(cero_df)

Viajes con duracion negativa: 0
Viajes con duracion cero: 6
Viajes con duracion menor a 1 minuto: 72
Viajes con duracion mayor a 3 horas: 0

Viajes con duración cero
                  pickup             dropoff  distance  fare  total
1690 2019-03-22 06:24:14 2019-03-22 06:24:14       0.0  72.0   72.0
5493 2019-03-06 12:14:00 2019-03-06 12:14:00       0.0  10.0   10.0
5624 2019-03-11 14:04:50 2019-03-11 14:04:50       0.0   2.5    3.3
5638 2019-03-13 12:22:00 2019-03-13 12:22:00       0.0  10.0   10.0
6083 2019-03-19 15:34:00 2019-03-19 15:34:00       0.0   5.0    5.0
6356 2019-03-01 16:58:23 2019-03-01 16:58:23       0.0   3.0    4.8


### 2.3 Velocidad promedio (6 pts)

La velocidad promedio de un viaje, en millas por hora, es `distance / (duracion_min / 60)`.

1. Crea la columna `velocidad_mph` **solo para viajes con duración mayor que cero**; en los demás debe quedar como faltante (`NaN`). Sugerencia: calcula la velocidad para todos los viajes y conserva solo los valores válidos con `.where(condición)`, que deja `NaN` donde la condición es `False`.
2. Cuenta los viajes con velocidad mayor a 60 mph (velocidad poco plausible en tráfico urbano).
3. Muestra los 5 viajes con mayor velocidad, ordenados de mayor a menor, con `pickup`, `dropoff`, `duracion_min`, `distance`, `velocidad_mph` y `fare`.

In [46]:
# Escribe tu código aquí\
# calcular 'velocidad_mph' solo para viajes con duracion > 0
taxis['velocidad_mph'] = (taxis['distance'] / (taxis['duracion_min'] / 60)).where(
    taxis['duracion_min'] > 0
)

#contar viajes con velocidad > 60 mph
mayor_60 = (taxis['velocidad_mph'] > 60).sum()
print("Viajes con velocidad mayor a 60 mph:", mayor_60)

#tpo 5 de viajes con mayor velocidad
columnas = ['pickup', 'dropoff', 'duracion_min', 'distance', 'velocidad_mph', 'fare']
top5 = taxis.nlargest(5, 'velocidad_mph')[columnas]
print()
print("Top 5 viajes con mayor velocidad:")
print(top5)

Viajes con velocidad mayor a 60 mph: 11

Top 5 viajes con mayor velocidad:
                  pickup             dropoff  duracion_min  distance  \
770  2019-03-02 03:16:59 2019-03-02 03:17:06      0.116667      9.40   
6089 2019-03-17 23:16:34 2019-03-17 23:16:52      0.300000     10.10   
491  2019-03-07 07:11:33 2019-03-07 07:11:39      0.100000      1.60   
5284 2019-03-22 06:24:02 2019-03-22 06:24:11      0.150000      1.70   
3241 2019-03-24 17:19:27 2019-03-24 17:19:29      0.033333      0.34   

      velocidad_mph  fare  
770     4834.285714   2.5  
6089    2020.000000  12.0  
491      960.000000   2.5  
5284     680.000000   2.5  
3241     612.000000  65.0  


### 2.4 Criterio de exclusión (7 pts)

Con base en 1.2, 2.2 y 2.3, define qué viajes excluirías del análisis y por qué. Considera:

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

b) ¿Qué umbrales (duración, velocidad, periodo) usarás? Justifícalos.

In [47]:
cortos = taxis[taxis['duracion_min'] < 1]

print("Total de viajes < 1 min:", len(cortos))
print()

# Estadsticas de distancia y tarifa para estos viajes
print("Distancia")
print(cortos['distance'].describe())
print()

print("Tarifa")
print(cortos['fare'].describe())
print()


Total de viajes < 1 min: 72

Distancia
count    72.000000
mean      0.447500
std       1.641412
min       0.000000
25%       0.000000
50%       0.000000
75%       0.102500
max      10.100000
Name: distance, dtype: float64

Tarifa
count     72.000000
mean      15.827222
std       24.742039
min        2.500000
25%        2.500000
50%        2.500000
75%       14.000000
max      120.000000
Name: fare, dtype: float64



**Respuesta 2.4:**

a) No todos son errores, unas son probables cancelaciones y otros viajes cortos

b)
* Umbral de velocidad ≤ 60 mph: descarta los casos de 10.1 millas en menos de  1 min
* Umbral de distancia > 0: descarta los 36 viajes con distance == 0
* Umbral de tarifa coherente: descartar fare < 2.5 o fare exageradamente alto para la distancia recorrida

---
## Actividad 3. Componentes y patrones (25 puntos)

### 3.1 Extraer componentes (6 pts)

Crea las columnas:

- `hora`: hora de `pickup`.
- `dia_num`: día de la semana de `pickup` (lunes = 0).
- `dia_nombre`: nombre del día de `pickup`.
- `fin_semana`: `True` si es sábado o domingo.
- `fecha`: solo la fecha (sin hora) de `pickup`.
- `cruza_medianoche`: `True` si la **fecha** de `dropoff` es distinta de la de `pickup`.

Cuenta cuántos viajes cruzan la medianoche.

In [48]:
# Escribe tu código aquí
#hora de pickup
taxis['hora'] = taxis['pickup'].dt.hour

#dia de la semana de pickup (lunes = 0)
taxis['dia_num'] = taxis['pickup'].dt.dayofweek

#nombre del dia de pickup
taxis['dia_nombre'] = taxis['pickup'].dt.day_name()

#fin de semana
taxis['fin_semana'] = taxis['dia_num'].isin([5, 6])

#solo la fecha sin hora de pickup
taxis['fecha'] = taxis['pickup'].dt.date

#fecha de dropoff distinta de la de pickup
taxis['cruza_medianoche'] = taxis['pickup'].dt.date != taxis['dropoff'].dt.date

#contar viajes que cruzan medianoche
cruzan = taxis['cruza_medianoche'].sum()
print("Viajes que cruzan la medianoche:", cruzan)
print()
print(taxis)

Viajes que cruzan la medianoche: 50

                  pickup             dropoff  passengers  distance  fare  \
0    2019-03-23 20:21:09 2019-03-23 20:27:24           1      1.60   7.0   
1    2019-03-04 16:11:55 2019-03-04 16:19:00           1      0.79   5.0   
2    2019-03-27 17:53:01 2019-03-27 18:00:25           1      1.37   7.5   
3    2019-03-10 01:23:59 2019-03-10 01:49:51           1      7.70  27.0   
4    2019-03-30 13:27:42 2019-03-30 13:37:14           3      2.16   9.0   
...                  ...                 ...         ...       ...   ...   
6428 2019-03-31 09:51:53 2019-03-31 09:55:27           1      0.75   4.5   
6429 2019-03-31 17:38:00 2019-03-31 18:34:23           1     18.74  58.0   
6430 2019-03-23 22:55:18 2019-03-23 23:14:25           1      4.14  16.0   
6431 2019-03-04 10:09:25 2019-03-04 10:14:29           1      1.12   6.0   
6432 2019-03-13 19:31:22 2019-03-13 19:48:02           1      3.85  15.0   

       tip  tolls  total   color      payment  ...

### 3.2 Patrón por hora del día (7 pts)

Con `groupby('hora')`, calcula para cada hora: el **número de viajes** y la **mediana** de `duracion_min`. Indica la hora con más viajes y la hora con menos.

Usa la mediana (no la media) porque la duración tiene valores extremos.

In [49]:
# Escribe tu código aquí
#agrupar por hora y calcular numero de viajes y mediana de duracion
por_hora = taxis.groupby('hora').agg(
    num_viajes=('duracion_min', 'size'),
    mediana_duracion=('duracion_min', 'median')
)

print("Patron por hora del dia:")
print(por_hora)
print()

#hora con mas viajes
hora_max = por_hora['num_viajes'].idxmax()
print("Hora con mas viajes:", str(hora_max).zfill(2) + ":00", por_hora.loc[hora_max, 'num_viajes'], "viajes")

# Hora con menos viajes
hora_min = por_hora['num_viajes'].idxmin()
print("Hora con menos viajes:", str(hora_min).zfill(2) + ":00",por_hora.loc[hora_min, 'num_viajes'], "viajes")

Patron por hora del dia:
      num_viajes  mediana_duracion
hora                              
0            205         10.733333
1            110          9.633333
2            101          9.183333
3             67          9.583333
4             57         10.500000
5             51          8.700000
6            142          7.833333
7            221          9.583333
8            313         11.316667
9            319         11.616667
10           325         11.883333
11           295         12.033333
12           334         12.775000
13           316         10.558333
14           360         11.841667
15           330         12.058333
16           336         11.583333
17           388         11.975000
18           417         11.100000
19           406         10.650000
20           368          9.258333
21           355         11.083333
22           321          9.816667
23           296         10.575000

Hora con mas viajes: 18:00 417 viajes
Hora con menos viajes: 05:

### 3.3 Patrón por día de la semana (6 pts)

Calcula por día de la semana el **número de viajes** y la **mediana** de `duracion_min` y de `fare`. Presenta los días en orden de lunes a domingo (sugerencia: agrupa por `['dia_num', 'dia_nombre']`).

Toma en cuenta que marzo de 2019 no tiene el mismo número de cada día de la semana: calcula también el **promedio de viajes por fecha**, dividiendo entre el número de fechas distintas de ese día de la semana (usa la columna `fecha` y `.nunique()`).

In [50]:
# Escribe tu código aquí

#agrupacion principal
por_dia = taxis.groupby(['dia_num', 'dia_nombre']).agg(
    num_viajes=('duracion_min', 'size'),
    mediana_duracion=('duracion_min', 'median'),
    mediana_fare=('fare', 'median')
)

#numero de fechas distintas por dia de la semana
fechas_por_dia = taxis.groupby(['dia_num', 'dia_nombre'])['fecha'].nunique()

#anadir columna de promedio
por_dia['fechas_distintas'] = fechas_por_dia
por_dia['promedio_viajes_por_fecha'] = (por_dia['num_viajes'] / por_dia['fechas_distintas']).round(2)

print("Patron por dia de la semana")
print(por_dia)

Patron por dia de la semana
                    num_viajes  mediana_duracion  mediana_fare  \
dia_num dia_nombre                                               
0       Monday             708         10.541667           9.5   
1       Tuesday            825         11.216667           9.5   
2       Wednesday          966         11.291667           9.5   
3       Thursday           905         11.766667           9.5   
4       Friday            1115         11.150000           9.5   
5       Saturday          1046         10.466667           9.0   
6       Sunday             868          9.983333           9.0   

                    fechas_distintas  promedio_viajes_por_fecha  
dia_num dia_nombre                                               
0       Monday                     4                     177.00  
1       Tuesday                    4                     206.25  
2       Wednesday                  4                     241.50  
3       Thursday                   5           

### 3.4 Interpretación (6 pts)

a) ¿Qué patrones observas por hora y por día? ¿Coinciden la hora con más viajes y la hora con mayor duración mediana? ¿Qué podría explicarlo?

b) ¿Por qué fue necesario calcular el promedio de viajes por fecha en 3.3 en lugar de comparar solo el total por día de la semana?

**Respuesta 3.4:**

a) La hora con más viajes es las 18:00 con 417 viajes, y la hora con menos viajes es las 05:00 con 51 viajes lo que correspondria con la logica de pensar que es debido a la demanda por horario laboral, pero en dado caso, la demanda por dia deberia reflejarlo de Lunes a Viernes y no es asi.
La hora pico (18:00) no coincide con la hora de mayor duración mediana (12:00), ni la hora valle (05:00) coincide con la de menor duración (06:00), podria explicarse que los viajes a la hora de salida de horario laboral son cortos probablemente para trasbordar y los viajes con mayor duracion son actividades ocasionales.

b)

---
## Actividad 4. Dataset limpio y comparación (20 puntos)

### 4.1 Dataset limpio (6 pts)

Crea `taxis_limpio` aplicando el criterio que definiste en 2.4. Imprime cuántos viajes se excluyeron por cada condición y en total, y verifica que `taxis_limpio` cumple todas las condiciones.

In [51]:
# Escribe tu código aquí
#definir cada condición por separado
cond_periodo    = (taxis['pickup'].dt.year == 2019) & (taxis['pickup'].dt.month == 3)
cond_duracion   = taxis['duracion_min'] >= 1
cond_velocidad  = (taxis['velocidad_mph'] > 0) & (taxis['velocidad_mph'] <= 60)
cond_distancia  = taxis['distance'] > 0

#contar exclusiones por condicion
print("Viajes excluidos por condicion")
print("Fuera de marzo 2019        :", (~cond_periodo).sum())
print("Duración < 1 minuto        :", (~cond_duracion).sum())
print("Velocidad fuera de (0, 60] :", (~cond_velocidad).sum())
print("Distancia <= 0             :", (~cond_distancia).sum())
print()

#mascara combinada
mascara = cond_periodo & cond_duracion & cond_velocidad & cond_distancia

#crear taxis_limpio
taxis_limpio = taxis[mascara].copy()

print("Total de viajes originales :", len(taxis))
print("Total de viajes excluidos  :", (~mascara).sum())
print("Total de viajes en limpio  :", len(taxis_limpio))
print()

#verificacion
print("Verificacion")
print("Todos en marzo 2019        :", ((taxis_limpio['pickup'].dt.year == 2019) & (taxis_limpio['pickup'].dt.month == 3)).all())
print("Todos con duracion >= 1    :", (taxis_limpio['duracion_min'] >= 1).all())
print("Todos con velocidad en (0,60]:", ((taxis_limpio['velocidad_mph'] > 0) & (taxis_limpio['velocidad_mph'] <= 60)).all())
print("Todos con distance > 0     :", (taxis_limpio['distance'] > 0).all())

Viajes excluidos por condicion
Fuera de marzo 2019        : 1
Duración < 1 minuto        : 72
Velocidad fuera de (0, 60] : 62
Distancia <= 0             : 51

Total de viajes originales : 6433
Total de viajes excluidos  : 85
Total de viajes en limpio  : 6348

Verificacion
Todos en marzo 2019        : True
Todos con duracion >= 1    : True
Todos con velocidad en (0,60]: True
Todos con distance > 0     : True


### 4.2 Comparación entre periodos (6 pts)

Con `taxis_limpio`, compara los viajes que inician entre las **7:00 y las 9:59** en **días hábiles** contra los del **mismo horario en fin de semana**. Para cada grupo reporta: número de viajes, mediana de `duracion_min` y mediana de `velocidad_mph`.

In [52]:
# Escribe tu código aquí
#filtrar horario 7-9
taxis_horario = taxis_limpio[taxis_limpio['hora'].between(7, 9)]

#agrupar por fin_semana
resumen = taxis_horario.groupby('fin_semana').agg(
    num_viajes=('duracion_min', 'size'),
    mediana_duracion=('duracion_min', 'median'),
    mediana_velocidad=('velocidad_mph', 'median')
)

print(resumen)

            num_viajes  mediana_duracion  mediana_velocidad
fin_semana                                                 
False              682         12.050000           9.106366
True               161          7.766667          13.473054


### 4.3 Registro de decisión (8 pts)

Redacta el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Conversión**: formato utilizado y verificación.
- **Zona horaria**: en qué zona están los datos y cómo lo verificaste.
- **Criterio de exclusión**: condiciones, umbrales y justificación.
- **Resultado**: viajes antes, después y excluidos.
- **Hallazgo de 4.2** y **limitaciones** del análisis.

**Respuesta 4.3:**

1. Conversión: Se usó pd.to_datetime() con formato '%Y-%m-%d %H:%M:%S'. Los faltantes antes y después coincidieron, no se generaron NaT.

2. Zona horaria: Los datos están en UTC. Se verificó filtrando el 10 de marzo de 2019 (día del DST en NY): 0 viajes a las 2:00 y 0 NaT al localizar con America/New_York. Si estuvieran en hora local NY, habría registros imposibles en ese rango.

3. Criterio de exclusión: (a) pickup en marzo 2019; (b) duración entre 1 y 180 min; (c) distancia > 0; (d) velocidad entre 0 y 60 mph. Justificados por los hallazgos de 1.2, 2.2 y 2.3.

4. Resultado: De [N_total] viajes originales se excluyeron [N_excl] ([P]%), quedando [N_limpio] en taxis_limpio. Verificado con .all().

5. Hallazgo 4.2: En el horario 7:00–9:59, los días hábiles tienen más viajes, mayor duración mediana y menor velocidad mediana que el fin de semana. El mismo horario tiene dinámicas distintas según el día.

6. Limitaciones: datos en UTC (no hora local NY), un solo mes (marzo 2019), desbalance de días, sin datos externos de tráfico/clima, conclusiones descriptivas no causales.

---
## Reflexión final (10 puntos)

1. ¿Qué ventajas tuvo indicar el formato explícito en 1.1 frente a dejar que pandas lo infiera? ¿En qué situación la inferencia habría producido errores silenciosos?
2. Supón que `pickup` se hubiera registrado en UTC y `dropoff` en hora local de Nueva York. ¿Qué efecto tendría en las duraciones calculadas? ¿Cómo lo habrías detectado con las herramientas de esta actividad?
3. Si se entrenara un modelo para predecir la duración de un viaje, ¿qué problemas causarían las anomalías temporales que encontraste si no se limpiaran?

**Respuesta final:**

1. Evita errores silenciosos de ambigüedad (día/mes, formatos mixtos)
2. Con las duraciones negativas, se detecta con describe() y conteo de negativas
3. Causarian ruido y métricas infladas

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.